### PR TITLE
Fix FPs about generic args

### PR DESCRIPTION
**Pull request by ThibsG**
Fix 2 false positives in [`use_self`] and [`useless_conversion`] lints, by taking into account generic args and comparing them.

Fixes: #7205 
Fixes: #7206 

changelog: Fix FPs about generic args in [`use_self`] and [`useless_conversion`] lints

### CODE REVIEW

File: tests/ui/use_self.rs [hunk 1]
```diff
@@ -462,3 +462,13 @@ mod issue6818 {
         a: i32,
     }
 }
+
+mod issue7206 {
+    struct MyStruct<const C: char>;
+
+    impl From<MyStruct<'a'>> for MyStruct<'b'> {
+        fn from(_s: MyStruct<'a'>) -> Self {
```
Comment (flip1995) [inline tests/ui/use_self.rs:470]: Wait, how doesn't this fail our test suite?

```suggestion
    impl From<MyStruct<'a>> for MyStruct<'b> {
        fn from(_s: MyStruct<'a>) -> Self {
```

Same in the other test file changes.

Comment (flip1995) [reply tests/ui/use_self.rs:470]: Oh nevermind, those are const generics as `char`s. Those are valid Rust code now. I have to get used to them I guess. :sweat_smile:

**Comment by mikerite** [comment]
It looks like these `use_self` warnings from serde_json seem to be missing after your change.

```
---> serde_json-1.0.64/src/de.rs:77:9
warning: unnecessary structure name repetition
  --> src/de.rs:77:9
   |
77 |         Deserializer::new(read::SliceRead::new(bytes))
   |         ^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self


---> serde_json-1.0.64/src/de.rs:84:9
warning: unnecessary structure name repetition
  --> src/de.rs:84:9
   |
84 |         Deserializer::new(read::StrRead::new(s))
   |         ^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self
```

I'm pretty sure they aren't false positives. I haven't made a minimal example. The types are quite complicated.

**Comment by mikerite** [comment]
```
#![allow(dead_code)]
#![warn(clippy::use_self)]

struct S<'a> {
    inner: &'a str,
}

struct S2<T> {
    inner: T,
}

impl<T> S2<T> {
    fn new() -> Self {
        unimplemented!();
    }
}

impl<'a> S2<S<'a>> {
    fn new_again() -> Self {
        S2::new()
    }
}

fn main() {}
```

**Comment by ThibsG** [comment]
Thanks @mikerite , nice catch. The comparison was too broad (not only for `Const` generic args).

Btw, where did you spot the missing warning? Locally? In the CI?

File: clippy_lints/src/useless_conversion.rs [hunk 2]
```diff
@@ -67,7 +67,7 @@ impl<'tcx> LateLintPass<'tcx> for UselessConversion {
                 if match_trait_method(cx, e, &paths::INTO) && &*name.ident.as_str() == "into" {
                     let a = cx.typeck_results().expr_ty(e);
                     let b = cx.typeck_results().expr_ty(&args[0]);
-                    if TyS::same_type(a, b) {
+                    if TyS::same_type(a, b) && !adt_different_const_generic_args(&a, &b) {
```
Comment (camsteffen) [inline clippy_lints/src/useless_conversion.rs:70]: `TyS::same_type` is *almost* what we need. So I think we should have our own version of it that includes consts comparisons.

```suggestion
                    if same_type_and_consts(a, b) {
```

File: clippy_utils/src/ty.rs [hunk 3]
```diff
@@ -322,3 +322,22 @@ pub fn walk_ptrs_ty_depth(ty: Ty<'_>) -> (Ty<'_>, usize) {
...
+        (&ty::Adt(did_a, substs_a), &ty::Adt(did_b, substs_b)) => {
+            if did_a != did_b {
+                return false;
+            }
+
+            substs_a
+                .types()
+                .zip(substs_b.types())
+                .all(|(a, b)| same_type_and_consts(a, b))
+                && substs_a.consts().zip(substs_b.consts()).all(|(a, b)| a == b)
```
Comment (camsteffen) [inline clippy_utils/src/ty.rs:339]: It should be faster to iterate the generic args and use `unpack`. The `types` and `consts` methods will iterate and unpack each list an extra time.

**Review by camsteffen — APPROVED** [review]
Looks good! Please squash commits.

**Comment by camsteffen** [comment]
Also nice photo 😀

**Comment by ThibsG** [comment]
Haha yeah, meet Léon, always happy to pull funny faces 😸

**Comment by camsteffen** [comment]
@bors r+

Thanks!

**Pull request merged**